### ISSUE TITLE
join_all() is too picky about lifetimes

### ISSUE THREAD

**Issue by dwrensha** [rust-lang-nursery/futures-rs#285]
Consider the following function:

```rust
fn foo<'a>(bufs: Vec<&'a [u8]>) -> Box<::futures::Future<Item=Vec<usize>, Error=()> + 'static> {
    let iter = bufs.into_iter().map(|b| ::futures::future::ok::<usize, ()>(b.len()));
    Box::new(::futures::future::join_all(iter))
}
```

I expect this to typecheck successfully, but I get an error:

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
  --> src/main.rs:35:21
   |
35 |     let iter = bufs.into_iter().map(|b| ::futures::future::ok::<usize, ()>(b.len()));
   |                     ^^^^^^^^^
   |
   = note: first, the lifetime cannot outlive the lifetime 'a as defined on unknown free region bounded by scope CodeExtent(212/CallSiteScope { fn_id: NodeId(122), body_id: NodeId(276) })...
note: ...so that types are compatible (expected std::vec::Vec<&[u8]>, found std::vec::Vec<&'a [u8]>)
  --> src/main.rs:35:21
   |
35 |     let iter = bufs.into_iter().map(|b| ::futures::future::ok::<usize, ()>(b.len()));
   |                     ^^^^^^^^^
   = note: but, the lifetime must be valid for the static lifetime...
note: ...so that the type `futures::JoinAll<std::iter::Map<std::vec::IntoIter<&[u8]>, [closure@src/main.rs:35:37: 35:84]>>` will meet its required lifetime bounds
  --> src/main.rs:36:5
   |
36 |     Box::new(::futures::future::join_all(iter))
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

The problem seems to be that `JoinAll<I>` is parameterized by the type of the `IntoIterator` that is passed into `join_all()`. In the above case, that type is not `'static`, even though its `Item` is `'static`.

**Comment by carllerche** [comment]
Is there an action item to take from this? The issue has been inactive for a year.

**Comment by stuhood** [comment]
@carllerche : I believe this will be useful for `0.2`, but requires changing the parameters of `JoinAll` see the linked PR and the other issue.

**Comment by alexcrichton** [comment]
@carllerche the fix, a breaking change, is at https://github.com/alexcrichton/futures-rs/pull/286 (linked above)

**Comment by stuhood** [comment]
Both this and #286 should be closable now, as they've landed in the `0.2` branch.

**Comment by alexcrichton** [comment]
Thanks @stuhood!
